### PR TITLE
Created Category API

### LINKED ISSUE #358: [Backend] - API creation & Updates to support Category Changes
**Description** 
Superuser should be able to add/modify categories and subcategories in the system. Updates must also be made to existing APIs to properly link a subcategory with its parent category

This includes
**Tasks**
- [x] Task 1: Update subcategories table with Category reference so each subcategory is part of a category
- [x] Task 2: Update IDatasetModel to now return a categoryID instead of category name
- [x] Task 3: Update /category to return ICategory Model
- [x] Task 4: Create POST /api/v1/Category API to add a new category & respective sub-category into the system
- [x] Task 5: Create PUT /api/v1/Category API to update a 
- [x] Task 6: Delete /api/v1/Category API to delete a category and ALL respective subcategories

interface ICategory {
name: string
id: number
subCategories: subCategories[]
}
interface ISubCategory {
name: string
id: number
}

**Acceptance Criteria**
- ICategory Model must be returned on a call to /fetchAllCategories
- Superuser must be able to add a category and subcategory
- Superuser must be able to modify a category and subcategory

### PR DESCRIPTION
**Pull request by silent-po2**
This PR creates a category API allowing for GET, PUT, POST, and DELETE operations for this field. It also includes other changes required in the backend, namely changes to data set upload/edit and data set search, in addition to controller tests which have near complete code coverage.

### CODE REVIEW

File: Server/src/controllers/CategoryController.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                let requestResponse: any = await this.categoryService.processAddCategory(categoryInfo);
+                return response.status(requestResponse.statusCode).json(requestResponse.message);
+            } catch (error) {
+                this.handleError(response, error);
+            }
+        }
+    }
+
+    private validateCreateCategoryRequest(request: Request): boolean {
+        return !request.body.name;
```
Comment (GD-Dheer) [inline Server/src/controllers/CategoryController.ts:31]: Maybe include the !request.body && !request.body.name

File: Server/src/controllers/CategoryController.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            return response.status(requestResponse.statusCode).json(requestResponse.message)
+        } catch (error) {
+            this.handleError(response, error);
+        }
+    }
+
+    async updateCategory(request: Request, response: Response): Promise<Response> {
+        let invalidResponse = this.validateUpdateCategoryRequest(request);
+        if (invalidResponse) {
+            return response.status(400).json("Request is invalid. Missing attributes")
```
Comment (GD-Dheer) [inline Server/src/controllers/CategoryController.ts]: What is missing here? Can provide a better error message

File: Server/src/models/CategoryModel.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+    * This method will return all existing dimensions
+    */
+    async getAllCategories(): Promise<ICategory[]> {
+        let categories = await Category.find();
+        let subcategories = await Subcategory.find();
+        let categoryModels = categories.map(category => {
+            let filteredSubcategories = subcategories.filter(value => value.categoryId == category.id)
```
Comment (GD-Dheer) [inline Server/src/models/CategoryModel.ts:43]: A+

File: Server/src/services/CategoryService.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    private requestResponse: IResponse = {} as any;
+    private categoryModel: CategoryModel
+
+    constructor() {
+        this.categoryModel = new CategoryModel()
+    }
+
+    async processAddCategory(categoryInfo: ICategory): Promise<IResponse> {
+        let name: boolean
+        name = await this.categoryModel.verifyIfNameExists(categoryInfo.name);
```
Comment (GD-Dheer) [inline Server/src/services/CategoryService.ts]: verifyIfCategoryExists?

File: Server/src/tests/controllers/CategoryController.spec.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    test('Invalid Update Category Request; delete in-use subcategory', async () => {
+        mockRequest = {
+            body: invalidUpdateData1
+        }
+        await controller.updateCategory(mockRequest as Request, mockResponse as Response)
+        expect(mockResponse.json).toBeCalledWith("Can't remove a subcategory as it in use by one or more data sets");
+        expect(mockResponse.status).toBeCalledWith(400);
+    });
+
+    test('Valid Update Category Request; no subcategory deletion', async () => {
```
Comment (GD-Dheer) [inline Server/src/tests/controllers/CategoryController.spec.ts:85]: A+ on tests

**Review by GD-Dheer — APPROVED** [review]
No real comments. Looks great. Good job on including tests too

File: Server/src/services/CategoryService.ts (hunk 4 above)
Comment (silent-po2) [reply Server/src/services/CategoryService.ts]: That's a good change. I'll do that!

File: Server/src/controllers/CategoryController.ts (hunk 1 above)
Comment (silent-po2) [reply Server/src/controllers/CategoryController.ts:31]: I'm not sure what the addition would do, as if a body is empty it is evidently missing the name. Explain?

File: Server/src/controllers/CategoryController.ts (hunk 2 above)
Comment (silent-po2) [reply Server/src/controllers/CategoryController.ts]: I'll add that to the error message

File: Server/src/controllers/CategoryController.ts (hunk 1 above)
Comment (ds161833) [reply Server/src/controllers/CategoryController.ts:31]: this is basically a null check for body, if you try to access name on body and body is null, you get NullReferenceException, you're exactly right, if body is null, JS will short circuit and will not try to get the name -> no null exception if body is null

**Review by ds161833 — APPROVED** [review]
nice pr, a lot of holes patched, good job on the tests, but im sad you didnt use the factories for test data :(

**Pull request merged**